cni/cmd/install/main.go

```diff
@@ -0,0 +1,316 @@
+		if err != nil {
+			return err
+		}
+		logger.Default.Debug("resulting config", zap.String("config", string(marshaled)))
+
+		err = atomic.WriteFile(mountedCniNetDir+"/"+resolvedName, bytes.NewReader(marshaled))
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+	return nil
+}
+
+func transformJsonConfig(kumaCniConfig string, hostCniConfig []byte) ([]byte, error) {
+	queryString := `if has("type") then
+   .plugins = [.]
+   | del(.plugins[0].cniVersion)
+   | to_entries
+   | map(select(.key=="plugins"))
+   | from_entries
+   | .plugins += [` + kumaCniConfig + `]
+   | .name = "k8s-pod-network"
+   | .cniVersion = "0.3.0"
+else
+  del(.plugins[]? | select(.type == "kuma-cni"))
+  | .plugins += [` + kumaCniConfig + `]
+end`
+	// this is probably going to be rewritten to not use `jq` at all
+	query, err := gojq.Parse(queryString)
+	if err != nil {
+		return nil, err
+	}
+
+	var parsed map[string]interface{}
+	err = json.Unmarshal(hostCniConfig, &parsed)
+	if err != nil {
+		return nil, err
+	}
+
+	result := query.Run(parsed)
+	modified, _ := result.Next()
+	marshaled, err := json.MarshalIndent(modified, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+	return marshaled, nil
+}
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return err == nil
+}
+
+func prepareKubeconfig(mountedCniNetDir, kubecfgName, serviceAccountPath string) error {
+	serviceAccountTokenPath := serviceAccountPath + "/token"
+	serviceAccountToken, err := ioutil.ReadFile(serviceAccountTokenPath)
+	if err != nil {
+		return err
+	}
+
+	if fileExists(serviceAccountTokenPath) {
+		kubernetesServiceHost := env.GetString("KUBERNETES_SERVICE_HOST", "")
+		if kubernetesServiceHost == "" {
+			logger.Default.Error("KUBERNETES_SERVICE_HOST env variable not set")
+			os.Exit(1)
+		}
+
+		kubernetesServicePort := env.GetString("KUBERNETES_SERVICE_PORT", "")
+		if kubernetesServicePort == "" {
+			logger.Default.Error("KUBERNETES_SERVICE_PORT env variable not set")
+			os.Exit(1)
+		}
+
+		kubeCaFile := env.GetString("KUBE_CA_FILE", serviceAccountPath+"/ca.crt")
+		kubeCa, err := ioutil.ReadFile(kubeCaFile)
+		if err != nil {
+			return err
+		}
+		kubernetesServiceProtocol := env.GetString("KUBERNETES_SERVICE_PROTOCOL", "https")
+		tlsConfig := "certificate-authority-data: " + base64.StdEncoding.EncodeToString(kubeCa)
+
+		kubeconfig := kubeconfigTemplate(kubernetesServiceProtocol, kubernetesServiceHost, kubernetesServicePort, string(serviceAccountToken), tlsConfig)
+		err = atomic.WriteFile(mountedCniNetDir+"/"+kubecfgName, strings.NewReader(kubeconfig))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func kubeconfigTemplate(protocol, host, port, token, tlsConfig string) string {
+	return `
+# Kubeconfig file for kuma CNI plugin.
+apiVersion: v1
+kind: Config
+clusters:
+- name: local
+  cluster:
+    server: ` + protocol + `://` + host + `:` + port + `
+    ` + tlsConfig + `
```

Comment:
You shouldn't build the url like this. It causes problems with ipv6 for example

Reply:
```
	serverUrl := url.URL{
		Scheme: ic.KubernetesServiceProtocol,
		Host:   net.JoinHostPort(safeHost, ic.KubernetesServicePort),
	}
```
and you can drop ifs with govalidators etc.